corehq/apps/hqadmin/couch_domain_utils.py

```diff
@@ -0,0 +1,148 @@
+    """
+    end = end_date or datetime.datetime.utcnow()
+    start = start_date or (end - datetime.timedelta(days=5))
+
+    couch_domains = domains or ACTIVE_COUCH_DOMAINS.get_enabled_domains()
+
+    for domain in couch_domains:
+        form_ids, has_discrepancies = _get_all_form_ids(domain, start, end)
+        if has_discrepancies:
+            metrics_gauge(
+                'commcare.es.couch_domain.couch_discrepancy_detected',
+                1,
+                tags={
+                    'domain': domain,
+                }
+            )
+            if stdout:
+                stdout.write(f"\nFound discrepancies in form counts for domain {domain}")
+        forms_not_in_es = _get_forms_not_in_es(form_ids)
+        if forms_not_in_es:
+            metrics_gauge(
+                'commcare.es.couch_domain.stale_docs_in_es',
+                len(forms_not_in_es),
+                tags={
+                    'domain': domain,
+                }
+            )
+            if stdout:
+                stdout.write(f"\nFound {len(forms_not_in_es)} forms not in es "
+                             f"for {domain}")
+            changes = _get_changes(domain, forms_not_in_es)
+            form_es_processor = get_xform_pillow().processors[0]
+            for change in changes:
+                form_es_processor.process_change(change)
+
+
+def _get_couch_form_ids(domain, start, end):
+    "Get form IDs from the 'by_domain_doc_type_date/view' couch view"
+    form_ids = set()
+    for doc_type in ['XFormInstance']:
+        startkey = [domain, doc_type, json_format_datetime(start)]
+        endkey = [domain, doc_type, json_format_datetime(end)]
+        results = XFormInstance.get_db().view(
+            'by_domain_doc_type_date/view', startkey=startkey, endkey=endkey,
+            reduce=False, include_docs=False
+        )
+        form_ids.update({row['id'] for row in results})
+    return form_ids
+
+
+def _get_all_form_ids(domain, start, end):
+    """Get form IDs from couch multiple times to ensure we get them all
+    This should not be necessary but for some reason we are seeing partial results.
+    Also returns a boolean in the second parameter to indicate if there were
+    any discrepancies in the results.
+    """
+    all_form_ids = None
+    previous_form_ids = None
+    form_id_differences = []
+    for i in range(10):
+        form_ids = _get_couch_form_ids(domain, start, end)
+        if not all_form_ids:
+            all_form_ids = form_ids
+        else:
+            form_id_differences.append(previous_form_ids.difference(form_ids))
+            all_form_ids |= form_ids
+        previous_form_ids = form_ids
+    diff_counts = [len(diff) for diff in form_id_differences]
+    return all_form_ids, any(diff_counts)
+
+
+def _get_forms_in_es(form_ids):
+    return (
+        FormES(
+            es_instance_alias=ES_EXPORT_INSTANCE
+        ).remove_default_filters().form_ids(form_ids).values_list('_id', flat=True)
+    )
```

Comment:
It'd be nice to use an `es.mget` query here instead, since I believe that's faster when doing a primary key lookup.  Probably not a blocker, though.